indico/core/db/sqlalchemy/protection.py

```diff
@@ -125,12 +129,212 @@ def can_access(self, user, acl_attr='acl', legacy_method='canAccess', allow_admi
         else:
             # should never happen, but since this is a sensitive area
             # we better fail loudly if we have garbage
             raise ValueError('Invalid protection mode: {}'.format(self.protection_mode))
+
+    def update_principal(self, principal, read_access=None):
+        """Updates access privileges for the given principal.
+
+        :param principal: A `User`, `GroupProxy` or `EmailPrincipal` instance.
+        :param read_access: If the principal should have explicit read
+                            access to the object.
+        :return: The ACL entry for the given principal or ``None`` if
+                 he was removed (or not added).
+        """
+        principal = _resolve_principal(principal)
+        principal_class, entry = _get_acl_data(self, principal)
+        if entry is None and read_access:
+            entry = principal_class(principal=principal)
+            self.acl_entries.add(entry)
+            signals.acl.entry_changed.send(type(self), obj=self, principal=principal, entry=entry, is_new=True,
+                                           old_data=None)
+            return entry
+        elif entry is not None and not read_access:
+            self.acl_entries.remove(entry)
+            signals.acl.entry_changed.send(type(self), obj=self, principal=principal, entry=None, is_new=False,
+                                           old_data=None)
+            return None
+        return entry
+
+    def remove_principal(self, principal):
+        """Revokes all access privileges for the given principal.
+
+        This method doesn't do anything if the user is not in the
+        object's ACL.
+
+        :param principal: A `User`, `GroupProxy` or `EmailPrincipal` instance.
+        """
+        principal = _resolve_principal(principal)
+        entry = _get_acl_data(self, principal)[1]
+        if entry is not None:
+            signals.acl.entry_changed.send(type(self), obj=self, principal=principal, entry=None, is_new=False,
+                                           old_data=entry.current_data)
+            self.acl_entries.remove(entry)
+
+
+class ProtectionManagersMixin(ProtectionMixin):
+    @memoize_request
+    def can_manage(self, user, role=None, allow_admin=True, check_parent=True, explicit=False):
+        """Checks if the user can manage the object.
+
+        :param user: The :class:`.User` to check. May be None if the
+                     user is not logged in.
+        :param: role: The management role that is needed for the
+                      check to succeed.  If not specified, full
+                      management privs are required.  May be set to
+                      the string ``'ANY'`` to check if the user has
+                      any management privileges.  Full management
+                      privs always grant access even if the role is
+                      not granted explicitly.
```

Comment:
`event.can_manage(user, 'registration')` will succeed if the guy has `full_access=True` on his ACL entry, even if he doesn't have the 'registration' role. Basically `full_access` is like superuser, unless `explicit` is specified.